### PR TITLE
[201911][Mellanox] Collect MST dump before syncd restart on shutdown notification

### PR DESCRIPTION
**Pull request by vivekrnv**
#### Why I did it

Collecting MST dump before syncd restart on shutdown notification during a SAI failure

Dump can be found under:
```
root@sonic:/home/admin# ls -l /var/dump/mstdump/
total 10684
-rw-r--r-- 1 root root 5460332 Aug 15 18:41 mstdump_20220815_184143.tar.gz
-rw-r--r-- 1 root root 5473253 Aug 15 21:46 mstdump_20220815_214642.tar.gz

root@sonic:/home/admin# tar -xvzf /var/dump/mstdump/mstdump_20220815_214642.tar.gz
├── ir-gdb
│   └── core
└── mstdump
    ├── mstdump1
    ├── mstdump2
    ├── mstdump3
    └── mststatus
```

#### How I did it

Checked for shutdown notification log in sairedis and used it to determine whether the shutdown is normal or due to SAI failure

#### How to verify it

Simulated a SAI failure event and verified it. Verified it also on different reboots and config reload scenarios the dump is not generated

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205

### CODE REVIEW

**Review by liat-grozovik — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@qiluo-msft please review.

**Review by dgsudharsan — APPROVED** [review]

**Comment by liat-grozovik** [comment]
@qiluo-msft @prsunny any objection to merge it to 201911?

File: files/scripts/syncd.sh [hunk 1]
```diff
@@ -136,6 +136,41 @@ wait() {
     /usr/bin/${SERVICE}.sh wait $DEV
 }
 
+collect_mst() {
+    debug "Collecting MST dump before syncd restart"
+    IRGDBSERVER=/usr/local/bin/ir-gdbserver
```
Comment (qiluo-msft) [inline files/scripts/syncd.sh:141]: >ir-gdbserver [](http://example.com/codeflow?start=31&length=12)

Not installed package?

File: files/scripts/syncd.sh [hunk 2]
```diff
@@ -154,7 +189,12 @@ stop() {
         /bin/systemctl stop pmon
         debug "Stopped pmon service"
     fi
-
+    if [[ x$sonic_asic_platform == x"mellanox" ]]; then
+        # In case of SAI failure the last line of sairedis.rec would contain switch_shutdown_request
+        if tail -1 /var/log/swss/sairedis.rec | grep -q switch_shutdown_request; then
+            collect_mst
```
Comment (qiluo-msft) [inline files/scripts/syncd.sh:195]: >collect_mst [](http://example.com/codeflow?start=12&length=11)

Should this PR be raised on master first, and they backport to all the branches (>= 201911)?

Comment (qiluo-msft) [reply files/scripts/syncd.sh:188]: Still applicable.

Comment (dgsudharsan) [reply files/scripts/syncd.sh:188]: @qiluo-msft This change is not required for master. In master we have an event driven firmware tool which will automatically create these dumps on any firmware failure event.

Comment (qiluo-msft) [reply files/scripts/syncd.sh:188]: How about branches between (201911, master)

**Review by qiluo-msft — APPROVED** [review]

**Comment by liat-grozovik** [comment]
From Mellanox pov it is critical for 201911 for now
As for other branches we are checking if this is needed as there is a mother mechanism to collect dumps which cannot be backported to 201911

Let's ensure we have it in 201911 and if needed we will have a PR against other branches if needed

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

File: files/scripts/syncd.sh (hunk 2 above)
Comment (dgsudharsan) [reply files/scripts/syncd.sh:188]: It exists from 202012 and so we are good.

**Comment by prgeor** [comment]
@liat-grozovik this is now included in 20191130.88

**Pull request merged**